File: scripts/build-windows-installer.mjs
import { copyFileSync, existsSync, mkdtempSync, readFileSync, readdirSync, rmSync } from 'node:fs';
import { join } from 'node:path';
import { tmpdir } from 'node:os';

const projectRoot = process.cwd();
const stableRoot = join(projectRoot, 'build', 'stable-win-x64', 'webwaifu3-electrobun');
const outputDir = join(projectRoot, 'build', 'stable-win-x64');
const resourcesDir = join(stableRoot, 'Resources');
const installerScriptPath = join(projectRoot, 'installer', 'windows', 'desktopwaifu.iss');
const packageJsonPath = join(projectRoot, 'package.json');

function fail(message) {
	console.error(`[windows-installer] ${message}`);
	process.exit(1);
}

function resolveVersion() {
	const tag = process.env.DESKTOPWAIFU_VERSION ?? process.env.GITHUB_REF_NAME ?? '';
	if (tag) {
		return tag.startsWith('v') ? tag.slice(1) : tag;
	}

	if (!existsSync(packageJsonPath)) {
		return '0.0.1';
	}

	const pkg = JSON.parse(readFileSync(packageJsonPath, 'utf8'));
	return pkg.version || '0.0.1';
}

function resolveIsccPath() {
	const candidates = [
		process.env.INNO_SETUP_COMPILER,
		join(process.env.LOCALAPPDATA || '', 'Programs', 'Inno Setup 6', 'ISCC.exe'),
		'C:\\Program Files (x86)\\Inno Setup 6\\ISCC.exe',
		'C:\\Program Files\\Inno Setup 6\\ISCC.exe',
	].filter(Boolean);

	return candidates.find((candidate) => existsSync(candidate));
}

function extractStablePayload() {
	if (!existsSync(resourcesDir)) {
		fail(`Stable resources directory not found at ${resourcesDir}`);
	}

	const archiveName = readdirSync(resourcesDir).find((entry) => entry.endsWith('.tar.zst'));
	if (!archiveName) {
		fail(`No stable payload archive found in ${resourcesDir}`);
	}

	const archivePath = join(resourcesDir, archiveName);
	const tempRoot = mkdtempSync(join(tmpdir(), 'desktopwaifu-installer-'));
	const proc = Bun.spawnSync(['tar', '-xf', archivePath, '-C', tempRoot], {
		cwd: projectRoot,
		stdio: ['ignore', 'inherit', 'inherit'],
	});

	if (proc.exitCode !== 0) {
		rmSync(tempRoot, { recursive: true, force: true });
		fail(`Failed to extract stable payload archive ${archivePath}`);
	}

	const extractedRoots = readdirSync(tempRoot, { withFileTypes: true }).filter((entry) => entry.isDirectory());
	if (extractedRoots.length !== 1) {
		rmSync(tempRoot, { recursive: true, force: true });
		fail(`Expected exactly one extracted app root in ${tempRoot}`);
	}

	return {
		tempRoot,
		sourceDir: join(tempRoot, extractedRoots[0].name),
	};
}

if (!existsSync(stableRoot)) {
	fail(`Stable bundle not found at ${stableRoot}. Run bun run build:stable first.`);
}

const isccPath = resolveIsccPath();
if (!isccPath) {
	fail('Inno Setup compiler not found. Install Inno Setup 6 or set INNO_SETUP_COMPILER.');
}

const version = resolveVersion();
const outputBaseFilename = 'DesktopWaifu-Setup';
const extracted = extractStablePayload();
const extractedLauncherPath = join(extracted.sourceDir, 'bin', 'launcher');
const extractedLauncherExePath = join(extracted.sourceDir, 'bin', 'launcher.exe');

if (!existsSync(extractedLauncherExePath) && existsSync(extractedLauncherPath)) {
	copyFileSync(extractedLauncherPath, extractedLauncherExePath);
	console.log('[windows-installer] Copied extracted launcher to launcher.exe for shortcut/uninstall integration');
}

if (!existsSync(extractedLauncherExePath)) {
	rmSync(extracted.tempRoot, { recursive: true, force: true });
	fail(`launcher.exe not found in extracted bundle ${join(extracted.sourceDir, 'bin')}`);
}

const proc = Bun.spawnSync(
	[
		isccPath,
		`/DMyAppVersion=${version}`,
		`/DSourceDir=${extracted.sourceDir}`,
		`/DOutputDir=${outputDir}`,
		`/DOutputBaseFilename=${outputBaseFilename}`,
		installerScriptPath,
	],
	{
		cwd: projectRoot,
		stdio: ['ignore', 'inherit', 'inherit'],
	},
);

if (proc.exitCode !== 0) {
	rmSync(extracted.tempRoot, { recursive: true, force: true });
	fail(`Inno Setup failed with exit code ${proc.exitCode}`);
}

rmSync(extracted.tempRoot, { recursive: true, force: true });
console.log(`[windows-installer] Built ${join(outputDir, `${outputBaseFilename}.exe`)}`);

File: genie-tts/build_windows_installer.py
from __future__ import annotations

import argparse
import json
import os
import subprocess
from datetime import datetime
from pathlib import Path

from build_windows_onedir import main as build_onedir

ROOT = Path(__file__).resolve().parent
DIST_SOURCE_DIR = ROOT / "build" / "windows-onedir" / "DesktopWaifu-GenieCompat"
INSTALLER_OUTPUT_DIR = ROOT / "build" / "windows-installer"
INSTALLER_SCRIPT_PATH = ROOT / "installer" / "windows" / "desktopwaifu-genie-compat.iss"
ROOT_PACKAGE_JSON_PATH = ROOT.parent / "package.json"
LOCAL_PACKAGE_JSON_PATH = ROOT / "package.json"


def fail(message: str) -> None:
    raise SystemExit(f"[genie-installer] {message}")


def resolve_version() -> str:
    tag = str(os.environ.get("DESKTOPWAIFU_VERSION") or os.environ.get("GITHUB_REF_NAME") or "").strip()
    if tag:
        return tag[1:] if tag.startswith("v") else tag

    for package_json_path in (ROOT_PACKAGE_JSON_PATH, LOCAL_PACKAGE_JSON_PATH):
        if not package_json_path.exists():
            continue
        payload = json.loads(package_json_path.read_text("utf-8"))
        version = str(payload.get("version") or "").strip()
        if version:
            return version

    return "0.0.1"


def resolve_iscc_path() -> Path:
    candidates = [
        os.environ.get("INNO_SETUP_COMPILER"),
        str(Path(os.environ.get("LOCALAPPDATA") or "") / "Programs" / "Inno Setup 6" / "ISCC.exe"),
        r"C:\Program Files (x86)\Inno Setup 6\ISCC.exe",
        r"C:\Program Files\Inno Setup 6\ISCC.exe",
    ]

    for candidate in candidates:
        if candidate and Path(candidate).exists():
            return Path(candidate)

    fail("Inno Setup compiler not found. Install Inno Setup 6 or set INNO_SETUP_COMPILER.")


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description="Build the DesktopWaifu Genie Windows installer.")
    parser.add_argument(
        "--rebuild-onedir",
        action="store_true",
        help="Rebuild the PyInstaller onedir payload before running Inno Setup.",
    )
    return parser.parse_args()


def main() -> None:
    args = parse_args()

    if args.rebuild_onedir or not DIST_SOURCE_DIR.exists():
        build_onedir()

    if not DIST_SOURCE_DIR.exists():
        fail(f"Compiled onedir payload not found at {DIST_SOURCE_DIR}")
    if not INSTALLER_SCRIPT_PATH.exists():
        fail(f"Inno Setup script not found at {INSTALLER_SCRIPT_PATH}")

    INSTALLER_OUTPUT_DIR.mkdir(parents=True, exist_ok=True)
    version = resolve_version()
    output_base_filename = "DesktopWaifu-GenieCompat-Setup"
    iscc_path = resolve_iscc_path()

    existing_installer = INSTALLER_OUTPUT_DIR / f"{output_base_filename}.exe"
    if existing_installer.exists():
        try:
            existing_installer.unlink()
        except PermissionError:
            timestamp_suffix = datetime.now().strftime("%Y%m%d-%H%M%S")
            output_base_filename = f"{output_base_filename}-{timestamp_suffix}"

    proc = subprocess.run(
        [
            str(iscc_path),
            f"/DMyAppVersion={version}",
            f"/DSourceDir={DIST_SOURCE_DIR}",
            f"/DOutputDir={INSTALLER_OUTPUT_DIR}",
            f"/DOutputBaseFilename={output_base_filename}",
            str(INSTALLER_SCRIPT_PATH),
        ],
        cwd=str(ROOT),
        check=False,
    )
    if proc.returncode != 0:
        fail(f"Inno Setup failed with exit code {proc.returncode}")

    built_installer = INSTALLER_OUTPUT_DIR / f"{output_base_filename}.exe"
    if not built_installer.exists():
        fail(f"Expected installer was not created at {built_installer}")

    print(f"Built DesktopWaifu Genie installer to: {built_installer}")


if __name__ == "__main__":
    main()
